File: java/_94_Binary_Tree_Inorder_Traversal.java
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class _94_Binary_Tree_Inorder_Traversal {
	public List<Integer> inorderTraversal(TreeNode root) {
        TreeNode tmp=root;
        Stack<TreeNode>stack=new Stack<>();
        List<Integer>ans=new ArrayList<>();
        while(tmp!=null || !stack.isEmpty()) {
        	while(tmp!=null) {
        		stack.push(tmp);
        		tmp=tmp.left;
        	}
        	if(!stack.isEmpty()) {
        		tmp=stack.pop();
        		ans.add(tmp.val);
        		tmp=tmp.right;
        	}
        }
        return ans;
    }
}

File: cpp/94.Binary Tree Inorder Traversal.cpp
#include <vector>
#include <stack>
using namespace std;


 struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 };

class Solution {
private:
    void dfs(TreeNode* root, vector<int> &ret){
        if (root==NULL)
            return;
        dfs(root->left, ret);
        ret.push_back(root->val);
        dfs(root->right, ret);
    }
public:
//    vector<int> inorderTraversal(TreeNode* root) {
//        vector<int> ret;
//        dfs(root, ret);
//        return ret;
//    }
    vector<int> inorderTraversal(TreeNode* root) {
        stack<TreeNode *>_stack;
        vector<int>ret;
        TreeNode* tmpNode = root;
        while(!_stack.empty() || tmpNode != NULL){
            while(tmpNode != NULL){
                _stack.push(tmpNode);
                tmpNode = tmpNode -> left;
            }
            tmpNode = _stack.top();
            _stack.pop();
            ret.push_back(tmpNode -> val);
            tmpNode = tmpNode ->right;
        }
        return ret;
    }
};